snuba/query/snql/parser.py

```diff
@@ -77,15 +84,16 @@
 
 snql_grammar = Grammar(
     r"""
-    query_exp             = match_clause select_clause group_by_clause? where_clause? having_clause? order_by_clause? limit_by_clause? limit_clause? offset_clause? granularity_clause? totals_clause? space*
+    query_exp             = match_clause select_clause group_by_clause? array_join_clause? where_clause? having_clause? order_by_clause? limit_by_clause? limit_clause? offset_clause? granularity_clause? totals_clause? space*
 
     match_clause          = space* "MATCH" space+ entity_match
     select_clause         = space+ "SELECT" space+ select_list
     group_by_clause       = space+ "BY" space+ group_list
+    array_join_clause     = space+ "ARRAY JOIN" space+ column_name
```

Comment:
There is an issue that needs to be addressed if we introduce the array join clause in SnQL. Clickhouse array join clause does not support aliases. 
See this thread for context: https://github.com/getsentry/snuba/pull/1486#discussion_r523239367

Which means we need to de-mangle them re-opening the door to alias shadowing. We went ahead with that solution in the PR above because SnQL did not have the arrayjoin clause. But if we introduce it we need to address the problem.

I can think about two possible solutions, but maybe there are more:
- Never use arrayjoin clause from clickhouse, and use calls to the arrayJoin function instead. It should be possible to always express the same query, but this needs to be verified.
- Allow mangling of the arrayjoin clause and ensure, during formatting that the allias is applied to the arrayjoin clause and used elsewhere (and not the other way around). 

I would prefer the first.

Reply:
Interesting. Although since the SnQL parser is using the same `_mangle_aliases` function, won't that function also unmangle the array join clause? Or is the idea to avoid that altogether?

Reply:
It would, but ideally we should not be unmangling. Unmangling reopens the door to alias shadowing unless we can always guarantee that the column used in the arrayjoin clause is a simple column, without adding an alias (document the reason clearly in the parser). 
We also need to ensure we invalidate queries that pass references to aliases in the arrayjoin clause (this should be a 400, it should not get to clickhouse).

Reply:
OK I changed this so SnQL will not have an array join clause at all. Users will have to supply any array join columns as a function. I tested this and the function does allow for aliases, so we don't have to validate that.